### PR TITLE
Object types design

### PR DESCRIPTION
**Pull request by Pazus**
Implementation was discussed in #23.
This was tested on 11.2.0.4 database.

This is a complete rewrite based on an object model to use features like inheritence, compositor pattern to create tree-like suites composed of suites, custom reporters and decorators for plugable enrichments based on inheritance of abstract class and interface.

A lot of work has to be done on enchancing assert capabilities, suite manager to automaticaly generate suits and test based on test-package methods. Though this might be a base of the framework.

As there is no naming convention defined for the project I tried to stay with the naming style used befor in v3 branch. But I thick naming convention has to be clarified in the future.

### CODE REVIEW

**Comment by Shoelace** [comment]
i dont see anythign wrong with this..  most of my 'suggesions' are more just additional developemnt.

the only thing here that i see different from the current (that i had worked on) is teh use of do_resolve and the setting `owner_name         => coalesce(trim(a_owner_name)                                                                          ,sys_context('userenv', 'current_schema'))`  in ut_test.tyb

in the previous version this was left blank if not specified and backfilled by do_resolve.  i did this becuase running `package.procedure;` is not the same as `user.package.procedure` if there is a synonym in place.

i assumed it would be best to leave it to  dbms_utility.name_resolve to fill the owner so the 'VALID' check works correctly.  i think the catch is it will need to be authid current_user to work..  but a lot of stuff will need to be changed on that front.

oh.. and i dont think its in teh proposed standards yet but i think most of us would prefer varible names and comments in english.  which include the git commit comments.

**Comment by Pazus** [comment]
If you like I can user sql formatter first and put settings file along other files of the project

**Comment by jgebal** [comment]
Putting code through formatter sounds like a good idea

**Comment by Pazus** [comment]
@Shoelace I use dbms_sql for dynamic SQL invocation as dbms_sql uses Current user rights to prevent from unauthorized access ([11.1 docs](https://docs.oracle.com/cd/B28359_01/appdev.111/b28419/d_sql.htm#i1027587)). There is not need for authid current_user with such approach.

**Comment by Shoelace** [comment]
I wasnt aware of the feature of dbms_sql. But it does make sense when you think about it. So cool. 1 less hassle for now. :D


_____________________________

**Comment by Pazus** [comment]
@Shoelace Sorry for russian commit comments in between I tought pull commit translates into single commit in the main project. Whatever I combined all temp commits into one so it's not a problem any more.

Also I have returned do_resolve behavior to the one you had and mad some additional changes

**Comment by jgebal** [comment]
I've ran all the examples on 11g XE.
All running and passing.
This is a major rework so I'd suggest merging this PR, so that others can contribute to the new codebase.
@Pazus , did you apply the formatting provided by @Shoelace ?

File: examples/RunExampleComplexSuiteWithDBMSOutputReporter.sql [hunk 1]
```diff
@@ -0,0 +1,45 @@
+--Shows how to create a test suite with the default reporter which is dbms_output
+--No tables are used for this.   
+--Suite Management packages are when developed will make this easier.
+Clear Screen
+Set Serveroutput On Size Unlimited
```
Comment (jgebal) [inline examples/RunExampleComplexSuiteWithDBMSOutputReporter.sql]: I would change the example scripts to make them self contained by:
- calling script to create example ut package(s) before execution of script
```sql
@ut_exampletest.pks
@ut_exampletest.pkb
@ut_exampletest2.pks
@ut_exampletest2.pkb
```
- drop the package after the script is finished (optionally)

**Comment by Pazus** [comment]
We will do that shortly, I'll create new pull request.

**Comment by Pazus** [comment]
I have merged PR from @jgebal which has some minor improvements. I hope now this PR is ready to be merged also.

**Comment by rlove** [comment]
I like the design changes.

**Pull request merged**